flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -539,26 +554,35 @@ public int tryFulfilledRequirementWithResource(
                 return numUnfulfilled;
             }
 
-            Queue<InternalResourceInfo> resourceInfoInUtilizationOrder =
-                    new PriorityQueue<>(
-                            internalResources.size(),
-                            Comparator.comparingDouble(i -> i.utilization));
-            resourceInfoInUtilizationOrder.addAll(internalResources);
+            Queue<InternalResourceInfo> prioritizedResources =
+                    new PriorityQueue<>(internalResources.size(), resourceInfoComparator);
+            prioritizedResources.addAll(internalResources);
 
-            while (numUnfulfilled > 0 && !resourceInfoInUtilizationOrder.isEmpty()) {
-                final InternalResourceInfo currentTaskManager =
-                        resourceInfoInUtilizationOrder.poll();
+            while (numUnfulfilled > 0 && !prioritizedResources.isEmpty()) {
+                final InternalResourceInfo currentTaskManager = prioritizedResources.poll();
 
                 if (currentTaskManager.tryAllocateSlotForJob(jobId, requiredResource)) {
                     numUnfulfilled--;
 
-                    // ignore non resource task managers to reduce the overhead of insert.
-                    if (!currentTaskManager.availableProfile.equals(ResourceProfile.ZERO)) {
-                        resourceInfoInUtilizationOrder.add(currentTaskManager);
+                    // ignore non-fitting task managers to reduce the overhead of insert and check.
+                    if (currentTaskManager.availableProfile.allFieldsNoLessThan(requiredResource)) {
+                        prioritizedResources.add(currentTaskManager);
                     }
                 }
             }
+
             return numUnfulfilled;
         }
+
+        public static PrioritizedResourceMatchingStrategy leastUtilization() {
```

Comment:
nit: Can we move these factory methods above the constructor and add proper JavaDoc (instead of just one comment) to both of the factory methods?